lib/kontena/actors/etcd_watcher.rb

```diff
@@ -0,0 +1,51 @@
+require 'concurrent/timer_task'
+require 'etcd'
+
+module Kontena::Actors
+  class EtcdWatcher < Concurrent::Actor::RestartingContext
+    include Kontena::Logging
+
+    attr_reader :client, :path
+
+    def initialize(host, path)
+      info "initialized with etcd host=#{host} path=#{path}"
+      @host = host
+      @path = path
+      @read_task = Concurrent::TimerTask.new(execution_interval: 1) {
+        read_etcd
+      }
+    end
+
+    def default_executor
+      Concurrent.global_io_executor
+    end
+
+    # @param [Message] msg
+    def on_message(msg)
+      case msg.action
+      when :start
+        start
+      else
+        pass
+      end
+    end
+
+    def start
+      @read_task.execute
+    end
+
+    def client
+      @client ||= Etcd::Client.new(host: @host, port: 2379)
+    end
+
+    def read_etcd
+      response = client.get(path, recursive: true)
+      self.parent << Message.new(:generate_config, response)
```

Comment:
This makes things async, right? This starts a chain of events: `generate_config` -> `write_config` -> `update_haproxy`. If some of the steps take a while could the order of updates be messed up? I mean later update going on the fast-lane and update first to haproxy.

Reply:
Yes, it's async but I don't see how it could mess up order of updates?

Reply:
Need to understand how the actors behave... do they process messages strictly in serial? If they can't start processing a new message until the previous message has been processed, then this isn't a problem.

Reply:
My thinking was something like:
`t` means time, "seconds".
t=0, etcd reader reads and emits the data
t=0.5, `generate_config` invocated (invocation a), for some reason it slows down
t=1, etcd reader reads and emits the data
t=1.5, `generate_config` invocated (invocation b)
t=1.7, `generate_config` invocated (invocation b) finishes and triggers `write_config`
t=1.8, `generate_config` invocated (invocation a) finishes and triggers `write_config`
...
vóila, ordering messed up :)